File: getting-started/python/binary_search.py
def search(data, key):
    start_idx = 0
    end_idx = len(data) - 1

    while start_idx <= end_idx:
        middle_idx = (start_idx + end_idx) // 2

        if data[middle_idx] == key:
            return middle_idx
        elif data[middle_idx] < key:
            start_idx = middle_idx + 1
        else:
            end_idx = middle_idx - 1

    return -1


if __name__ == "__main__":
    data = [i**2 for i in range(20)]

    print("--- data ---")
    print(data)

    print("input key > ", end="")
    key = int(input())

    result = search(data, key)
    if result != -1:
        print(f"result = {result}")
    else:
        print(f"{key} is not found.")

File: getting-started/golang/binary_search.go
package main

import (
	"fmt"
)

func search(data []int, key int) int {
	startIdx := 0
	endIdx := len(data) - 1

	for startIdx <= endIdx {
		middleIdx := (startIdx + endIdx) / 2

		if data[middleIdx] == key {
			return middleIdx
		} else if data[middleIdx] < key {
			startIdx = middleIdx + 1
		} else {
			endIdx = middleIdx - 1
		}
	}

	return -1
}

func main() {
	data := make([]int, 20)

	for i := range 20 {
		data[i] = i * i
	}

	fmt.Println("--- data ---")
	fmt.Println(data)

	fmt.Print("input key > ")
	var key int
	_, err := fmt.Scanf("%d", &key)
	if err != nil {
		fmt.Println("input error")

		return
	}

	result := search(data, key)
	if result != -1 {
		fmt.Printf("result = %d\n", result)
	} else {
		fmt.Printf("%d is not found.\n", key)
	}
}
